### ISSUE TITLE
is it possible to return null instead of nothing when using a nullable column in a table

### ISSUE THREAD

**Issue by kir4ik** [juanluispaz/ts-sql-query#84]
Hi,

I would like to get all requested fields and specify them in the return types. In order for Typescript to check that I did not forget some optional field to request

```ts
import {MySqlConnection} from 'ts-sql-query/connections/MySqlConnection';
import {MySqlPoolQueryRunner} from 'ts-sql-query/queryRunners/MySqlPoolQueryRunner';
import {Table} from 'ts-sql-query/Table';
import {SelectedValues} from 'ts-sql-query/extras/types';

class Connection extends MySqlConnection<'Connection'> {
  allowEmptyString = true;
}

const createMysqlConnection = () => new Connection(new MySqlPoolQueryRunner(mysqlPool, 'mySql'));

class OrderTable extends Table<Connection, 'OrderTable'> {
  id = this.autogeneratedPrimaryKey('id', 'int');
  name = this.column('name', 'string');
  rawField = this.optionalColumn('rawField', 'int');
  anotherRawField = this.optionalColumn('anotherRawField', 'string');

  constructor() {
    super('sa_order');
  }
}

const orderTable = new OrderTable();

const connection = createMysqlConnection();

type MyOrder = Pick<SelectedValues<OrderTable>, 'id' | 'name' | 'rawField'>;

const getOrders = async (): Promise<MyOrder[]> => {
  const selectQuery = connection.selectFrom(orderTable).select({
    id: orderTable.id,
    name: orderTable.name,
    /**
     * let's say I now want to work with the "rawField" field somewhere in the code,
     * I will add it to the types (MyOrder), but the TS does not tell me that I forgot to specify it here
     */
    // rawField: orderTable.rawField
  });
  return selectQuery.executeSelectMany();
}

(async () => {
  const [firstOrder] = await getOrders();
  firstOrder.rawField // undefined | number - ok, but connection.select does not select this field from the database, so it will never be there
})()
```

i think it would be better if the requested fields were explicitly specified as null, then TS can reporte that this field is not returned

because in this situation, when we see "undefined" by types, we cannot understand exactly what it means - the field was not requested or the field in the database is "null"

---
It seems to me that it would be better if only null was used, because undefined is not always obvious, either it turns to null on insert/update or this column will not be in the query. And with the select, either the field is really null or it was not specified in the query

**Comment by juanluispaz** [comment]
Hi,

null, undefined, and the presence or absence of the property with that values, it had been an intense debate in the JS & TS community; without a real consensus at the end; few versions ago, TypeScript introduced a [new flag](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-4.html#exact-optional-property-types---exactoptionalpropertytypes) to deal with this a little bit more. I didn't test the library when this flag was enabled, but it sounds like that will be something to keep in consideration regarding what you raise here.

What is modelled in ts-sql-query right now is any optional property that will be projected with an interrogation mark (in consequence, it allows undefined); this is a very common pattern in JS/TS community. GraphQL Codegen, for example, goes further, allowing null values as well in optional properties.

I think what you suggest here is to have the option to represent properties in objects as required nullable properties; in the code, that will be easy to implement, but in the TypeScript type inferences, it will be hard. Let me think about what I can imagine here, but it will require time, and I cannot promise it soon.

The issue with that approach will be if you combine it with any JSON representation, like Rest, where this is not the common way to proceed. As always, this debate never has a conclusive end. Let me know about your thought and how you will combine this with all the ecosystems in JS, including the browser side.

**Comment by kir4ik** [comment]
I think for my example above, the following variant of the result type from select is suitable

```ts
type SelectRowResult<T, U extends keyof T> = Pick<T, U> & {
  [K in Exclude<keyof T, U>]?: undefined
};

const getOrders = async (): Promise<MyOrder[]> => {
  const selectQuery = connection.selectFrom(orderTable).select({
    id: orderTable.id,
    name: orderTable.name,
  });
  /**
   * Type '{ id: number; name: string; }[]' is not assignable to type 'SelectRowResult<MyOrder, "id" | "name">[]'.
   * Type '{ id: number; name: string; }' is not assignable to type 'SelectRowResult<MyOrder, "id" | "name">'.
   * Property 'rawField' is missing in type '{ id: number; name: string; }' but required in type '{ rawField: undefined; }'.ts(2322)
   */
  const result: SelectRowResult<MyOrder, 'id' | 'name'>[] = await selectQuery.executeSelectMany();

  return result;
}
```

then the TS will report a problem in the code.
But it looks like a crutch.

**Comment by Smert** [comment]
I think that if a field was queried from a table, the result object must have this field. Now fields with `NULL` values don't exist in the result object. It can be unsafe because next we don't know is it is NULL or we forgot to request this field.

**Comment by juanluispaz** [comment]
Hi,

I'm exploring different possibilities here:

Having this table definition:

```ts
const tCustomer = new class TCustomer extends Table<DBConnection, 'TCustomer'> {
    id = this.autogeneratedPrimaryKey('id', 'int')
    firstName = this.column('first_name', 'string')
    lastName = this.column('last_name', 'string')
    birthday = this.optionalColumn('birthday', 'localDate')
    companyId = this.column('company_id', 'int')

    constructor() {
        super('customer'); // table name in the database
    }
}()
```

There are several cases to cover here:

1- **The output type coming from the result of a query (select, and in the case returning is used on insert, update or delete):**

```ts
type Output = SelectedValues<typeof tCustomer>
```

The generated type will be:

```ts
type Output = {
    id: number;
    firstName: string;
    lastName: string;
    companyId: number;
    birthday?: Date | undefined;
}
```

With your suggestion, the generated type will be:

 ```ts
type Output = {
    id: number;
    firstName: string;
    lastName: string;
    companyId: number;
    birthday: Date | null;
}
```

2- **The insert input type:**

```ts
type InsertInput = InsertableValues<typeof tCustomer>
```

The generated type will be:

```ts
type InsertInput = {
    firstName: string;
    lastName: string;
    companyId: number;
    id?: number | undefined;
    birthday?: Date | null | undefined;
}
```

Not sure what to do here because the absence of the value and the null or undefined values has a different meaning, and because of that, there are several [ingore](https://ts-sql-query.readthedocs.io/en/stable/supported-operations/#insert-definition) methods available when you build the query.

3- **The update input type:**

```ts
type UpdateInput = UpdatableValues<typeof tCustomer>
```

The generated type will be:

```ts
type UpdateInput = {
    firstName?: string | undefined;
    lastName?: string | undefined;
    companyId?: number | undefined;
    id?: number | undefined;
    birthday?: Date | null | undefined;
}
```

Not sure what to do here because the absence of the value and the null or undefined values has a different meaning,  and because of that, there are several [ingore](https://ts-sql-query.readthedocs.io/en/stable/supported-operations/#update-definition) methods available when you build the query.

4- **The update on insert conflict input type:**

```ts
type UpadetOnInsertConflictInput = UpdatableOnInsertConflictValues<typeof tCustomer>
```

The generated type will be:

```ts
type UpadetOnInsertConflictInput = {
    firstName?: string | undefined;
    lastName?: string | undefined;
    companyId?: number | undefined;
    id?: number | undefined;
    birthday?: Date | null | undefined;
}
```

Not sure what to do here because the absence of the value and the null or undefined values has a different meaning, and because of that, there are several [ingore](https://ts-sql-query.readthedocs.io/en/stable/supported-operations/#insert-definition) methods available when you build the query.

---

The output type is clear, but the input type is where things get messy.

I'm working on ways to change the insert/update/delete input object shape and as well limiting its available columns, something like:

```ts
    const value = {
        customerFirstName: 'John',
        customerLastName: 'Smith',
        customerCompanyId: 1
    }

    const insertedCustomerId = await connection.insertInto(tCustomer)
        .shapedAs({
            customerFirstName: tCustomer.firstName,
            customerLastName: tCustomer.lastName,
            customerCompanyId: tCustomer.companyId
        })
        .set(value)
        .returningLastInsertedId()
        .executeInsert()
```

What I'm thinking is to keep input types as they are right now. In that way, you have all the flexibility to control what you want, but, in the case the shape is provided, turn the type in the same way as the output type; in that way, the provided object must match the shape.

# What do you think?

---

There is an additional feature that can be affected: [select picking columns](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#select-dynamically-picking-columns).

In this case, I think the appropriate way to proceed is to keep the current behaviour of that property, marking them as optional (with ? in the property, allowing undefined), but returning null in the case they were picked. If that is not the desired behaviour, there is already available [Ignorable expression as null](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/#select-dynamically-picking-columns) that allows you to get the output as a null valued when you don't want to query it.

# Please, let me know what you think.

**Comment by mpricope** [comment]
Hey,

I might have a similar/related problem with this. .  So what I was trying to do is build a custom fragment that will return a value or null. 

Now if I use something like ```db.fragmentWithType("double", "optional")```, I can have my selected value be null, but it will just be absent in the result. 

I would really like to have it ```null``` in the result (otherwise I need to manually iterate on it and set it myself). 

On a more general note I think that on Insert/Update if you push a null/undefined the DB result is kinda the same (you get a NULL in your DB). So how the TypeScript shape is ... I don't know if it's that important. 

But for output/query I think it's important that ```null```'s are present and are very distinct from ```undefined```.

**Comment by juanluispaz** [comment]
Hi,

I've been thinking a lot about how to move forward on this coherently across the library.

There are two cases to keep in mind:

1. Input types (required by insert, update, etc.)
2. Output types (returned by a select, insert returning, update returning, delete returning)

I don't see how to create a coherent, useful model for this; then, I've been thinking, if instead of trying to put ts-sql-query running in a specific way, I offer you a variant of the select then you can pick the model you need for your specific use case. This will imply I will need to implement alternative versions to elements that return objects:

- `.select({...})` in select queries
- `.returning({...})` in insert, update, delete
- `.aggregateAsArray({...})` in the connection

I'm thinking about prefixing the names with the word `strict`; if someone has a better idea, let me know.

With this, your query will look like this:

```ts
const customerId = 10;

const customerWithId = connection.selectFrom(tCustomer)
   .where(tCustomer.id.equals(customerId))
   .strictSelect({
       id: tCustomer.id,
       firstName: tCustomer.firstName,
       lastName: tCustomer.lastName,
       birthday: tCustomer.birthday
   })
   .executeSelectOne();
```

The result type will be:
```tsx
const customerWithId: Promise<{
   id: number;
   firstName: string;
   lastName: string;
   birthday: Date | null;
}>
```

The inner object elimination in left joins will remain the same but returning null instead of undefined:

```ts
const parent = tCompany.forUseInLeftJoinAs('parent');

const companyWithParent = connection.selectFrom(tCompany)
   .leftJoin(parent).on(tCompany.parentId.equals(parent.id))
   .strictSelect({
       id: tCompany.id,
       name: tCompany.name,
       parent: {
           id: parent.id,
           name: parent.name,
       }
   })
   .executeSelectMany();
```

The result type will be:
```tsx
const companyWithParent: Promise<{
   id: number;
   name: string;
   parent: null | {
       id: number;
       name: string;
   };
}[]>
```

Implement this will require:

- Add `.strictSelect({...})` in select queries
- Add `.strictReturning({...})` in insert, update, delete
- Add `.strictAggregateAsArray({...})` in the connection
- Add `StrictSelectedValues` and `SelectedRow` to the extra types (advance usage)

The advantage of this is that you keep the freedom and don't see yourself in an everything-or-nothing situation, allowing you to have a migration path. The disadvantage is if you want always return nulls in your project, you must agree with your team to always use the alternative implementation.

This will keep input types as is now, allowing alternative implementation in the future where you are free to choose.

What do you think? Can I start implementing this?

Remember, I need your feedback on this.

**Comment by juanluispaz** [comment]
Additional things to consider:

Select picking columns will continue to produce optional columns, but if those values are optional, then, as well, can be null (when picked and the value is null)

Additional things to implement:

- Add `expandStrictTypeFromDynamicPickPaths` as an alternative to `expandTypeFromDynamicPickPaths` in `ts-sql-query/dynamicCondition`

**Comment by juanluispaz** [comment]
I'm thinking about changing the wording a little bit: use the word `tight` instead of `strict`

In that way, the changes will be:

- Add `.tightSelect({...})` in select queries
- Add `.tightReturning({...})` in insert, update, delete
- Add `.aggregateAsTightArray({...})` in the connection
- Add `TightSelectedValues` and `TightSelectedRow` to the extra types (advance usage)
- Add `expandTightTypeFromDynamicPickPaths` as an alternative to `expandTypeFromDynamicPickPaths` in `ts-sql-query/dynamicCondition`

I will start implementing this as an experimental feature.

Let me know your feedback.

**Comment by mpricope** [comment]
I think having the choice of behavior for the ```read``` part is just great.  Another idea for how to specify this, if it's possible in TS, would be something like:
```.select({}).strict().executeSelectOne()```

Basically chaining that after the main method. This way you just add this on the documentation, and not another 3-4 methods that might be confusing. 

Another way would be to add this to the field itself when you do the projection: ```tCompany.name.strict()```.  This will be more verbose but then allow you very fine granularity on the projection level (although I don't know if this is really useful). 

I don't know which one of these, if any are feasible in TypeScript if you want to preserve all type magic :).

**Comment by juanluispaz** [comment]
Hi,

I like your idea of a chained call, making it easy to document the feature or add support to another variant.

Regarding the possibility of specifying per field, that will be too much and make the complex projections rules so messy; I think that can be implemented outside of ts-sql-query in an easy type-safe way according to the desired complex rule.

With this approach, the changes will be:

- Add `.select({...}).asTightObject()` in select queries
- Add `.returning({...}).asTightObject()` in insert, update, delete
- Add `.aggregateAsArray({...}).asTightObject()` in the connection
- Add `TightSelectedValues` and `TightSelectedRow` to the extra types (advance usage)
- Add `expandTightTypeFromDynamicPickPaths` as an alternative to `expandTypeFromDynamicPickPaths` in `ts-sql-query/dynamicCondition`

For the last two, the chained approach will not be convenient.

What do you think?

**Comment by mpricope** [comment]
This sounds pretty good.  I am a big confused although about ```strict``/```tight``` naming :). Would ```passNull``` or something related to ```null``` won't be more relevant?  I mean are these methods do anything else?
I think you might be looking at this from the TypeScript perspective, but at the end of the day I am just writing a query in SQL and I use TypeScript .. . well for type :). But I am thinking in terms of DB and SQL. The whole thing that is so cool about ts-sql-query ... is that it does just this.

**Comment by juanluispaz** [comment]
What about use `.projectingOptionalValuesAsNullable()` instead of `.asTightObject()`

With this naming, the changes will be:

- Add `.select({...}). projectingOptionalValuesAsNullable()` in select queries
- Add `.returning({...}). projectingOptionalValuesAsNullable()` in insert, update, delete
- Add `.aggregateAsArray({...}). projectingOptionalValuesAsNullable()` in the connection
- Add `SelectedValuesProjectedAsNullable` and `SelectedRowProjectedAsNullable` to the extra types (advance usage)
- Add `expandTypeProjectedAsNullableFromDynamicPickPaths` as an alternative to `expandTypeFromDynamicPickPaths` in `ts-sql-query/dynamicCondition`

Remember, even using `.projectingOptionalValuesAsNullable()`, the columns in a select picking columns will remain optional in the object, indicating when they were not picked.

I like this more; it is self-explaining and focuses on the query, as you suggested.

What do you think?

**Comment by mpricope** [comment]
Looks great and pretty self descriptive.

**Comment by juanluispaz** [comment]
Additionally to the changes described before, the following changes were required as well:

- Add`PickValuesPathProjectedAsNullable` as an alternative to`PickValuesPath` in `ts-sql-query/dynamicCondition`
- Add`PickValuesPathWitAllPropertiesProjectedAsNullable` as an alternative to`PickValuesPathWitAllProperties` in `ts-sql-query/dynamicCondition`

I hope to release this tomorrow.

**Comment by juanluispaz** [comment]
Released ts-sql-query 1.55.0, implementing the changes mentioned here. This version supports projecting optional values in an object as nullable in the select, insert, update, delete and aggregate array output. This feature makes the optional property required as nullable in the projected value.

Please, test it and let me know your feedback.
